// @hyperlane-xyz/core/dist/factories/contracts/interfaces/avs/vendored/IAVSDirectory__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "avs",
                type: "address",
            },
            {
                indexed: false,
                internalType: "string",
                name: "metadataURI",
                type: "string",
            },
        ],
        name: "AVSMetadataURIUpdated",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "operator",
                type: "address",
            },
        ],
        name: "deregisterOperatorFromAVS",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "operator",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "bytes",
                        name: "signature",
                        type: "bytes",
                    },
                    {
                        internalType: "bytes32",
                        name: "salt",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint256",
                        name: "expiry",
                        type: "uint256",
                    },
                ],
                internalType: "struct ISignatureUtils.SignatureWithSaltAndExpiry",
                name: "operatorSignature",
                type: "tuple",
            },
        ],
        name: "registerOperatorToAVS",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "metadataURI",
                type: "string",
            },
        ],
        name: "updateAVSMetadataURI",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class IAVSDirectory__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
