// @hyperlane-xyz/core/dist/factories/contracts/interfaces/avs/vendored/IPaymentCoordinator__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                components: [
                    {
                        components: [
                            {
                                internalType: "contract IStrategy",
                                name: "strategy",
                                type: "address",
                            },
                            {
                                internalType: "uint96",
                                name: "multiplier",
                                type: "uint96",
                            },
                        ],
                        internalType: "struct IPaymentCoordinator.StrategyAndMultiplier[]",
                        name: "strategiesAndMultipliers",
                        type: "tuple[]",
                    },
                    {
                        internalType: "contract IERC20",
                        name: "token",
                        type: "address",
                    },
                    {
                        internalType: "uint256",
                        name: "amount",
                        type: "uint256",
                    },
                    {
                        internalType: "uint64",
                        name: "startTimestamp",
                        type: "uint64",
                    },
                    {
                        internalType: "uint64",
                        name: "duration",
                        type: "uint64",
                    },
                ],
                internalType: "struct IPaymentCoordinator.RangePayment[]",
                name: "rangePayments",
                type: "tuple[]",
            },
        ],
        name: "payForRange",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class IPaymentCoordinator__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
