// @hyperlane-xyz/sdk/dist/gas/adapters/serialization.js
import { PublicKey } from '@solana/web3.js';
import { SealevelAccountDataWrapper, getSealevelAccountDataSchema, } from '../../utils/sealevelSerialization.js';
// Should match https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/rust/sealevel/programs/hyperlane-sealevel-igp/src/accounts.rs#L24
export var SealevelInterchainGasPaymasterType;
(function (SealevelInterchainGasPaymasterType) {
    // An IGP with gas oracles and that receives lamports as payment.
    SealevelInterchainGasPaymasterType[SealevelInterchainGasPaymasterType["Igp"] = 0] = "Igp";
    // An overhead IGP that points to an inner IGP and imposes a gas overhead for each destination domain.
    SealevelInterchainGasPaymasterType[SealevelInterchainGasPaymasterType["OverheadIgp"] = 1] = "OverheadIgp";
})(SealevelInterchainGasPaymasterType || (SealevelInterchainGasPaymasterType = {}));
/**
 * IGP Config Borsh Schema
 */
// Config schema, e.g. for use in token data
export class SealevelInterchainGasPaymasterConfig {
    fields;
    program_id;
    program_id_pubkey;
    type;
    igp_account;
    igp_account_pub_key;
    constructor(fields) {
        this.fields = fields;
        Object.assign(this, fields);
        this.program_id_pubkey = new PublicKey(this.program_id);
        this.igp_account_pub_key = this.igp_account
            ? new PublicKey(this.igp_account)
            : undefined;
    }
}
export const SealevelInterchainGasPaymasterConfigSchema = {
    kind: 'struct',
    fields: [
        ['program_id', [32]],
        ['type', 'u8'],
        ['igp_account', [32]],
    ],
};
/**
 * IGP Program Data Borsh Schema
 */
// Should match https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/rust/sealevel/programs/hyperlane-sealevel-igp/src/accounts.rs#L91
export class SealevelOverheadIgpData {
    fields;
    /// The bump seed for this PDA.
    bump;
    /// The salt used to derive the overhead IGP PDA.
    salt;
    /// The owner of the overhead IGP.
    owner;
    owner_pub_key;
    /// The inner IGP account.
    inner;
    inner_pub_key;
    /// The gas overheads to impose on gas payments to each destination domain.
    gas_overheads;
    constructor(fields) {
        this.fields = fields;
        Object.assign(this, fields);
        this.owner_pub_key = this.owner ? new PublicKey(this.owner) : undefined;
        this.inner_pub_key = new PublicKey(this.inner);
    }
}
export const SealevelOverheadIgpDataSchema = new Map([
    [
        SealevelAccountDataWrapper,
        getSealevelAccountDataSchema(SealevelOverheadIgpData, [8]),
    ],
    [
        SealevelOverheadIgpData,
        {
            kind: 'struct',
            fields: [
                ['bump', 'u8'],
                ['salt', [32]],
                ['owner', { kind: 'option', type: [32] }],
                ['inner', [32]],
                ['gas_overheads', { kind: 'map', key: 'u32', value: 'u64' }],
            ],
        },
    ],
]);
