// @hyperlane-xyz/core/dist/factories/contracts/libs/StaticWeightedValidatorSetFactory__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [],
        name: "PACKAGE_VERSION",
        outputs: [
            {
                internalType: "string",
                name: "",
                type: "string",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "signingAddress",
                        type: "address",
                    },
                    {
                        internalType: "uint96",
                        name: "weight",
                        type: "uint96",
                    },
                ],
                internalType: "struct IStaticWeightedMultisigIsm.ValidatorInfo[]",
                name: "_validators",
                type: "tuple[]",
            },
            {
                internalType: "uint96",
                name: "_thresholdWeight",
                type: "uint96",
            },
        ],
        name: "deploy",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "address",
                        name: "signingAddress",
                        type: "address",
                    },
                    {
                        internalType: "uint96",
                        name: "weight",
                        type: "uint96",
                    },
                ],
                internalType: "struct IStaticWeightedMultisigIsm.ValidatorInfo[]",
                name: "_validators",
                type: "tuple[]",
            },
            {
                internalType: "uint96",
                name: "_thresholdWeight",
                type: "uint96",
            },
        ],
        name: "getAddress",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "implementation",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
export class StaticWeightedValidatorSetFactory__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
